manifests/compose.pp

```diff
@@ -7,118 +7,50 @@
-#
-# @param base_url
-#   The base url for installation
-#   This allows use of a mirror that follows the same layout as the
-#   official repository
-#
-# @param raw_url
-#   Override the raw URL for installation
-#   The default is to build a URL from baseurl. If rawurl is set, the caller is
-#   responsible for ensuring the URL points to the correct version and
-#   architecture.
-#
-# @param curl_ensure
-#   Whether or not the curl package is ensured by this module.
-#
 class docker::compose (
-  Enum[present,absent] $ensure                 = present,
-  Optional[String]               $version      = $docker::params::compose_version,
-  Optional[String]               $install_path = $docker::params::compose_install_path,
-  Optional[String]               $symlink_name = $docker::params::compose_symlink_name,
-  Optional[Pattern['^((http[s]?)?:\/\/)?([^:^@]+:[^:^@]+@|)([\da-z\.-]+)\.([\da-z\.]{2,6})(:[\d])?([\/\w \.-]*)*\/?$']] $proxy = undef,
-  Optional[String]               $base_url     = $docker::params::compose_base_url,
-  Optional[String]               $raw_url      = undef,
-  Optional[Boolean]              $curl_ensure  = $docker::params::curl_ensure,
-) inherits docker::params {
-  if $facts['os']['family'] == 'windows' {
-    $file_extension = '.exe'
-    $file_owner     = 'Administrator'
-  } else {
-    $file_extension = ''
-    $file_owner     = 'root'
-  }
-
-  $docker_compose_location           = "${install_path}/${symlink_name}${file_extension}"
-  $docker_compose_location_versioned = "${install_path}/docker-compose-${version}${file_extension}"
-
-  if $ensure == 'present' {
-    if $raw_url != undef {
-      $docker_compose_url = $raw_url
-    } else {
-      $docker_compose_url = "${base_url}/${version}/docker-compose-${facts['kernel']}-${facts['os']['hardware']}${file_extension}"
+  Enum[present,absent] $ensure  = present,
+  Optional[String]     $version = undef,
+) {
+  if $docker::manage_package {
+    include docker::params
+
+    $_version = $version ? {
+      undef   => $docker::params::compose_version,
+      default => $version,
     }
-
-    if $proxy != undef {
-      $proxy_opt = "--proxy ${proxy}"
+    if $_version and $ensure != 'absent' {
+      $package_ensure = $_version
     } else {
-      $proxy_opt = ''
+      $package_ensure = $ensure
     }
 
-    if $facts['os']['family'] == 'windows' {
-      $docker_download_command = "if (Invoke-WebRequest ${docker_compose_url} ${proxy_opt} -UseBasicParsing -OutFile \"${docker_compose_location_versioned}\") { exit 0 } else { exit 1}" # lint:ignore:140chars
-
-      $parameters = {
-        'proxy'                             => $proxy,
-        'docker_compose_url'                => $docker_compose_url,
-        'docker_compose_location_versioned' => $docker_compose_location_versioned,
+    case $facts['os']['family'] {
+      'Debian': {
+        $_require = $docker::use_upstream_package_source ? {
+          true  => Apt::Source['docker'],
```

Comment:
If I recall correctly, we must require `Class['apt::update']` otherwise Puppet can try to install the package before `apt update` is run and fail or install the wrong package.

Reply:
It looks like the [apt docs](https://forge.puppet.com/modules/puppetlabs/apt/reference#-apt--source--notify_update) have `notify_update` default to true on the `apt::source` resource.

Reply:
Yes, this is the other side of the requirements: by default, apt::source changes trigger an apt update, but the package installation can happen at any time, unless we say we want it to be after the apt update.

Similar case in another module: 

https://github.com/voxpupuli/puppet-gitlab/blob/master/manifests/omnibus_package_repository.pp#L58